pkg/threshcrypto/api.go

```diff
@@ -0,0 +1,35 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package threshcrypto
+
+// The ThreshCrypto interface represents an implementation of threshold cryptography primitives inside the MirModule module.
+// It is responsible for producing and verifying cryptographic threshold signatures, which disperses the authority
+// to sign among a group of N members, where T must sign their share for a full signature to be produced.
+// It internally stores information about the group's public key and the node's private key share.
+type ThreshCrypto interface {
+	// SignShare signs the provided data and returns the resulting signature share.
+	// The data to be signed is the concatenation of all the passed byte slices.
+	// A signature share produced by SignShare is verifiable using VerifyShare.
+	// After obtaining signature shares from T group members, the full signature can be constructed with Recover.
+	// Returns the signature (and a nil error) on success, and a non-nil error otherwise.
+	SignShare(data [][]byte) ([]byte, error)
+
+	// VerifyShare verifies that a signature share is valid for the given data.
+	// Returns nil on success (i.e., if the given signature share is valid) and a non-nil error otherwise.
+	VerifyShare(data [][]byte, signatureShare []byte) error
+
+	// Recover constructs a full signature from signature shares over data.
+	// The given array may contain invalid signature shares: as long as at least T distinct signature
+	// shares are valid, Recover will succeed, albeit less efficiently.
```

Comment:
I have mixed feelings regarding this property.
The implementation you are using does indeed verify the partial signatures in `Recover`.
However, it seems kind of redundant in my opinion as the shares can (and, probably, should) be verified before invoking `Recover`.

Maybe let's make it a property of the particular implementation and not of the abstraction (i.e., move it to the comment for `TBLSInst.Recover`)?

Reply:
In fact, it could be a serious performance overhead as verifying signature shares (i.e., doing pairings) is the most computationally heavy part. Not only this API forces the protocol to verify the shares multiple times (once before `Recover` and once inside), but it also prohibits parallelizing the verification.

I wonder why they did it this way.

Reply:
I agree, it's a good idea to force the optimizable design (verify shares, then recover) right from the start, even if some implementations are sub-optimal.

The implementation of `Recover` for tbls only uses public kyber APIs and is simple, so I think we can land the optimized implementation (without the double verification) here.